syntax = "proto3";

package market;

message RegisterBuyerRequest {
  string ip = 1;
  string port = 2;
}

message RegisterBuyerResponse {
  string status = 1;
}

message RegisterSellerRequest {
  string ip = 1;
  string port = 2;
  string uuid = 3;
}

message RegisterSellerResponse {
  string status = 1;
}

message SellItemRequest {
  string ip = 1;
  string port = 2;
  string uuid = 3;
  Item item = 4;
}

message SellItemResponse {
  string status = 1;
  int32 itemId = 2;
}

message UpdateItemRequest {
  string itemId = 1;
  float newPrice = 2;
  int32 newQuantity = 3;
  string ip = 4;
  string port = 5;
  string uuid = 6;
}

message UpdateItemResponse {
  string status = 1;
}

message DeleteItemRequest {
  string itemId = 1;
  string ip = 2;
  string port = 3;
  string uuid = 4;
}

message DeleteItemResponse {
  string status = 1;
}

message DisplaySellerItemsRequest {
  string ip = 1;
  string port = 2;
  string uuid = 3;
}

message DisplaySellerItemsResponse {
  repeated Item item = 1;
}

message Item {
  string itemId = 1;
  float price = 2;
  string name = 3;
  string category = 4;
  string description = 5;
  int32 quantity = 6;
  float rating = 7;
}

message SearchItemRequest {
  string name = 1;
  string category = 2;
}

message SearchItemResponse {
  repeated Item item = 1;
}

message BuyItemRequest {
  string itemId = 1;
  int32 quantity = 2;
  string ip = 3;
  string port = 4;
}

message BuyItemResponse {
  string status = 1;
}

message AddToWishListRequest {
  string itemId = 1;
  string ip = 2;
  string port = 3;
}

message AddToWishListResponse {
  string status = 1;
}

message RateItemRequest {
  string itemId = 1;
  string ip = 2;
  string port = 3;
  int32 rating = 4;
}

message RateItemResponse {
  string status = 1;
}

service MarketService {
  rpc RegisterBuyer(RegisterBuyerRequest) returns (RegisterBuyerResponse);
  rpc RegisterSeller(RegisterSellerRequest) returns (RegisterSellerResponse);
  rpc SellItem(SellItemRequest) returns (SellItemResponse);
  rpc UpdateItem(UpdateItemRequest) returns (UpdateItemResponse);
  rpc DeleteItem(DeleteItemRequest) returns (DeleteItemResponse);
  rpc DisplaySellerItems(DisplaySellerItemsRequest) returns (DisplaySellerItemsResponse);
  rpc SearchItem(SearchItemRequest) returns (SearchItemResponse);
  rpc BuyItem(BuyItemRequest) returns (BuyItemResponse);
  rpc AddToWishList(AddToWishListRequest) returns (AddToWishListResponse);
  rpc RateItem(RateItemRequest) returns (RateItemResponse);
}